src/hooks/project-hooks.tsx

```diff
@@ -79,27 +84,54 @@ export const ProjectProvider = ({
   const intl = useIntl();
   const toast = useToast();
   const logging = useLogging();
-  const setEditorOpen = useStore((s) => s.setEditorOpen);
   const projectEdited = useStore((s) => s.projectEdited);
   const expectChangedHeader = useStore((s) => s.setChangedHeaderExpected);
   const projectFlushedToEditor = useStore((s) => s.projectFlushedToEditor);
   const checkIfProjectNeedsFlush = useStore((s) => s.checkIfProjectNeedsFlush);
   const getCurrentProject = useStore((s) => s.getCurrentProject);
+  const navigate = useNavigate();
+  const doAfterEditorUpdatePromise = useRef<Promise<void>>();
   const doAfterEditorUpdate = useCallback(
     async (action: () => Promise<void>) => {
-      if (checkIfProjectNeedsFlush()) {
-        const project = getCurrentProject();
-        expectChangedHeader();
-        await driverRef.current!.importProject({ project });
-        projectFlushedToEditor();
+      if (!doAfterEditorUpdatePromise.current && checkIfProjectNeedsFlush()) {
+        doAfterEditorUpdatePromise.current = new Promise<void>(
+          (resolve, reject) => {
+            // driverRef.current is not defined on first render.
+            // Only an issue when navigating to code page directly.
+            if (!driverRef.current) {
+              reject(
+                new TypeError(
+                  "Cannot read properties of null (reading 'importProject')"
+                )
```

Comment:
This feels like it might be very confusing to future-us. Can we reject with something that's clearly custom?

Reply:
If we threw a custom error subclass we could check for it below.

Reply:
Otherwise happy with this.

Reply:
Added a `CodeEditorError` subclass with clearer message. Haven't checked for it though. See comment below. [5d0fab3](https://github.com/microbit-foundation/ml-trainer/pull/416/commits/5d0fab3ef2cea8eeea3799ea0b9d7fae2780c344)